### PR TITLE
feat: backlog wave 3 — compose in all bindings, honest introspection scope, race-free watch teardown, bench pipeline

### PR DESCRIPTION
**Pull request by baladithyab**
## Summary

Wave 3: 6 seeds + 1 carried-over review finding, three adversarially-reviewed parallel lanes (process as #57/#58).

### Compose through the bindings (rsac-fba7) — the big one
`Composition`/`CompositionBuilder`/`Group` exposed in all three bindings (+3,824 lines):
- **Python (PyO3)** & **Node (napi)** wrap the Rust compose API directly — same `RwLock` two-phase lock topology as rsac-8082, Python teardowns fully inside `allow_threads` (GIL rule), sync + async terminal-observable reads, context-manager/iterator protocols.
- **Go (cgo)** wraps the shipped `rsac_composition_*` C FFI; `librsac_ffi.a` now built `--features compose`.
- Hand-edited `index.d.ts` (+242) and `__init__.pyi` (+197) — no regeneration.
- Tests: python 15, napi 19, Go compose 11 (deadlock-topology stand-ins pin lock ordering).

### macOS honesty + memory (rsac-f547, rsac-e8aa / GH #32)
- `list_audio_applications_scoped()` + `#[non_exhaustive] ApplicationScope` — callers can finally tell *exact audio producers* from the *unfiltered fallback superset*; every fallback trigger point signals; CLI banner on fallback (verified live); the old `len>10 || Finder` test heuristic replaced with the honest API. Windows/Linux verified `ExactAudioProducers` honestly. Additive-only.
- Device-watch listeners moved to block-based `AudioObjectAddPropertyListenerBlock` on a self-owned serial queue; teardown removes on-queue then barriers → `WatchListenerContext` freed, not leaked (ends the ADR-0005 §5 intentional leak).

### CI / maintenance (rsac-75bf, rsac-1da3, rsac-62bb)
- Audio integration suite now triggers on stable release tags.
- `bench.yml`: weekly cron + dispatch criterion quick-mode A/B (with/without `bridge-zerocopy`), artifacts archived, fork-guarded; ADR-0006 promote-or-remove decision tracked in rsac-6508.
- Reference pins: cpal + screencapturekit-rs bumped, rtrb kept (matches shipped 0.3.4), docs reconciled.

### Carried-over (PR #57 outside-diff finding)
`__iter__` held the `iterating` mutex across `start()` (GIL-vs-lock circular-wait class) — fixed, `drop(iter_guard)` before start.

## Test plan
- [x] `bash scripts/gate.sh` (incl. docsrs replica) — OK
- [x] 635 lib tests; python 15/15; napi 19/19; Go compose 11/11
- [x] `rsac list-apps` live — fallback banner correct on this host
- [x] Watch teardown tests green (6 passed)
- [x] Adversarial review per lane: A + B MERGEABLE, C's fork-guard finding fixed in-lane
- [ ] CI matrix

Seeds: closes rsac-fba7, rsac-f547, rsac-e8aa (+ GH #32 comment), rsac-75bf, rsac-1da3, rsac-62bb on merge. New: rsac-6508 (ADR-0006 decision, blocked by bench data).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added multi-source audio composition APIs for Go, Node.js, and Python, including grouped sources, gain controls, mixing layouts, streaming, lifecycle management, and statistics.
  * Added scoped application enumeration with clear fallback reporting; the CLI now warns when results include all running applications.
* **Bug Fixes**
  * Improved macOS device-watcher teardown to prevent races and resource leaks.
  * Fixed Python iteration and composition shutdown edge cases.
* **Documentation**
  * Updated architecture, compatibility, changelog, and reference documentation.
* **Chores**
  * Added scheduled advisory benchmarks and stable-tag audio test runs.

### CODE REVIEW

File: bindings/rsac-python/src/lib.rs [hunk 1]
```diff
@@ -1495,6 +1502,771 @@ impl PyAudioCapture {
...
+            *iter_guard = true;
+
+            let guard = self_ref
+                .inner
+                .read()
+                .map_err(|e| PyRuntimeError::new_err(format!("Lock poisoned: {}", e)))?;
+            if let Some(ref comp) = *guard {
+                if !comp.is_running() {
+                    drop(guard);
+                    self_ref.start(py)?;
```
[the remark this answers is not in the document]
Comment (baladithyab) [reply bindings/rsac-python/src/lib.rs:2157]: ✅ fix-now: applied in `59a5ae0` — `Composition::__iter__` now drops `iter_guard` before `start()`, mirroring the `PyAudioCapture::__iter__` fix (`aaf678b`, same rsac-8082 GIL-vs-lock class). python tests 15/15.

File: src/core/introspection.rs [hunk 2]
```diff
@@ -196,32 +231,74 @@ pub fn list_audio_sources() -> AudioResult<Vec<AudioSource>> {
...
+            if is_fallback {
+                ApplicationScope::AllRunningFallback
+            } else {
+                ApplicationScope::ExactAudioProducers
+            }
         }
+        // Enumeration failed: push nothing and report an exact (empty) list —
+        // this is an honest "we found no audio producers", not an unfiltered
+        // superset. Errors are swallowed to keep discovery best-effort.
+        Err(_) => ApplicationScope::ExactAudioProducers,
```
[the remark this answers is not in the document]
Comment (baladithyab) [reply src/core/introspection.rs]: ✅ fix-now: applied in `59a5ae0` — new `ApplicationScope::EnumerationFailed` variant (additive; the enum is `#[non_exhaustive]`). All three platform arms now report it on a failed backend query instead of masquerading as an exact empty set; `rsac list-apps` prints an incomplete-list banner (unit-tested); CHANGELOG updated. Discovery stays best-effort — the error is still swallowed, but the scope is now honest.

File: bindings/rsac-go/compose.go [hunk 3]
```diff
@@ -0,0 +1,743 @@
...
+// AddSourceWithGain adds a capture source with an explicit linear mixdown gain
+// (1.0 = unity). The gain must be finite and >= 0; an invalid gain is rejected
+// eagerly with ErrInvalidParameter and the group is unchanged.
+func (g *Group) AddSourceWithGain(spec string, gain float32) error {
+	if g.handle == nil {
+		return ErrClosed
+	}
+	cspec := C.CString(spec)
+	defer C.free(unsafe.Pointer(cspec))
+	return newError(C.rsac_group_add_source_with_gain(g.handle, cspec, C.float(gain)))
```
[the remark this answers is not in the document]
Comment (baladithyab) [reply bindings/rsac-go/compose.go]: ✅ fix-now (`984eeb2`): `runtime.KeepAlive(g)`/`(b)` added after every cgo call on the finalizer-owned handles — Group setters/add-source, all builder setters, Preflight, and AddGroup (which keeps both b and g alive).

File: bindings/rsac-go/compose.go [hunk 4]
```diff
@@ -0,0 +1,743 @@
...
+		if err != nil {
+			if IsRecoverable(err) && !errors.Is(err, ErrClosed) {
+				runtime.Gosched()
+				continue
+			}
+			return // fatal terminal (or closed): stop delivering, close channel
+		}
+		if !ok {
+			runtime.Gosched()
+			continue
```
[the remark this answers is not in the document]
Comment (baladithyab) [reply bindings/rsac-go/compose.go:611]: ✅ fix-now (`984eeb2`): both stream loops now use `pollBackoff(ctx)` — a 1ms context-aware timer wait — for the empty and recoverable arms instead of `runtime.Gosched()`. One composed quantum is 10ms by default, so 1ms adds no perceptible latency while ending the busy-spin.

File: bindings/rsac-go/compose.go [hunk 5]
```diff
@@ -0,0 +1,743 @@
...
+	}
+}
+
+func (c *Composition) streamLoopWithErrors(ctx context.Context, ch chan<- StreamResult) {
+	defer close(ch)
+	for {
+		select {
+		case <-ctx.Done():
+			ch <- StreamResult{Err: ctx.Err()}
+			return
```
[the remark this answers is not in the document]
Comment (baladithyab) [reply bindings/rsac-go/compose.go:633]: ✅ fix-now (`984eeb2`): the cancellation report is now a non-blocking send (`select` with `default`) — best-effort delivery, and the goroutine always returns.

File: bindings/rsac-go/rsac.h [hunk 6]
```diff
@@ -526,6 +526,387 @@ uint32_t rsac_capabilities_supported_sample_rate_at(const RsacCapabilities* caps
...
+/**
+ * Point-in-time snapshot of a running composition's counters.
+ *
+ * Filled by rsac_composition_stats(). Plain value type — no heap, no free.
+ * Before rsac_composition_start() succeeds every field is 0.
+ */
+typedef struct {
+    uint64_t ticks;          /**< Composed buffers (ticks) emitted so far. */
+    uint64_t fallback_ticks; /**< Wall-clock fallback ticks (master stalled). */
+    size_t   num_sources;    /**< Composed sources; valid indices are 0..num_sources. */
```
[the remark this answers is not in the document]
Comment (baladithyab) [reply bindings/rsac-go/rsac.h]: ✅ fix-now (`984eeb2`): both curated headers, the FFI rustdoc, and the regenerated header now document the indices as half-open `[0, num_sources)`.

File: bindings/rsac-napi/src/lib.rs [hunk 7]
```diff
@@ -1056,6 +1056,594 @@ impl AudioCapture {
...
+                napi::Status::InvalidArg,
+                format!(
+                    "[ERR_RSAC_CONFIGURATION] gain {gain} is invalid (must be finite and >= 0)"
+                ),
+            ));
+        }
+        let target: rsac::CaptureTarget = spec.parse().map_err(audio_err_to_napi)?;
+        let mut g = self.inner.lock().map_err(lock_poisoned)?;
+        *g = g.clone().source_with_gain(target, gain as f32);
+        Ok(())
```
[the remark this answers is not in the document]
Comment (baladithyab) [reply bindings/rsac-napi/src/lib.rs:1118]: ✅ fix-now (`984eeb2`): both bindings validate AFTER narrowing to f32 and pass the validated `gain32` through; the predicate tests now include `f64::MAX` (finite as f64, +inf after narrowing) in the rejected set.

File: bindings/rsac-napi/src/lib.rs [hunk 8]
```diff
@@ -1056,6 +1056,594 @@ impl AudioCapture {
...
+    }
+
+    /// Remove the registered data callback and stop the pump. The `onEnd`
+    /// callback is left registered for a later session; use `offEnd` to clear it.
+    #[napi]
+    pub fn off_data(&self) -> Result<()> {
+        self.pump_active.store(false, Ordering::SeqCst);
+        let mut cb_guard = self.callback.lock().map_err(lock_poisoned)?;
+        *cb_guard = None;
+        Ok(())
```
[the remark this answers is not in the document]
Comment (baladithyab) [reply bindings/rsac-napi/src/lib.rs:1428]: ✅ fix-now (`984eeb2`) for the Composition pump in this PR's diff: a `pump_generation` AtomicU64 — each pump owns the generation current at spawn; `stop()`/`offData()` bump it, so a rapid re-`onData()` can't resurrect the old pump (it exits on generation mismatch) and a stale pump exiting only clears `pump_active` if it still owns the generation; spawn failure rolls the flag back. AudioCapture's identical PRE-EXISTING pump (outside this diff) is tracked: 📌 rsac-1a34.

File: src/audio/macos/coreaudio.rs [hunk 9]
```diff
@@ -1012,71 +1034,178 @@ pub(crate) unsafe fn remove_device_alive_listener(
...
+        //    every previously-enqueued notification block has finished (FIFO), so
+        //    when it returns no block is in flight or pending. This is the
+        //    in-flight barrier the old PROC API lacked.
+        queue.exec_sync(|| {});
+
+        // 3. Drop the block (releasing its captured `Arc` clone), then the
+        //    teardown's `Arc` clone → strong count hits 0 → context freed and its
+        //    `SyncSender` dropped, so the helper's recv() returns Err.
+        drop(block);
+        drop(context);
```
[the remark this answers is not in the document]
Comment (baladithyab) [reply src/audio/macos/coreaudio.rs]: ✅ fix-now (`984eeb2`): all three removal sites (WatchTeardown::run, add-registration rollback, helper-spawn rollback) now check the OSStatus. On any failure the code falls back to the pre-fix bounded leak for that watcher — `std::mem::forget` on block/context/queue and helper detach — so a still-registered listener's late dispatch can never touch freed memory; a warn! logs the fallback. The happy path (removal succeeds, never observed otherwise) keeps the freed-not-leaked behavior.

File: docs/designs/0005-device-watcher-raii-teardown.md [hunk 10]
```diff
@@ -3,19 +3,22 @@
...
-dispatch-queue rewrite is **deferred** (§5/§6). Pairs with ADR-0004.
+the in-flight-callback teardown race is now **RESOLVED** (rsac-e8aa / GH #32): the
+watcher uses **block-based** listeners (`AudioObjectAddPropertyListenerBlock`) on
+a **self-owned serial dispatch queue**, and teardown removes the block on that
+same queue then dispatches a **synchronous no-op barrier** which — by serial-queue
+FIFO ordering — cannot return until every in-flight block has finished, after
+which the `Arc<WatchListenerContext>` is dropped (**freed, not leaked**). This
+eliminates both the previously-intentional context leak and the race (§5). The
+device-*alive* listener (`DeviceAliveContext`) remains on the PROC API +
+`tearing_down` guard as a follow-up. Pairs with ADR-0004.
```
[the remark this answers is not in the document]
Comment (baladithyab) [reply docs/designs/0005-device-watcher-raii-teardown.md:27]: ✅ fix-now (`984eeb2`): ADR-0005 wording now qualifies the guarantee — freed-not-leaked on the expected path, with an explicit bounded-leak fallback confined to an OS-reported removal failure (matching the code from the sibling finding).

File: src/core/introspection.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (baladithyab) [reply src/core/introspection.rs]: ✅ fix-now (`5ff2203`+`59a5ae0`): new `ApplicationScope::EnumerationFailed` variant — all three platform arms report it on a failed backend query instead of an empty 'exact' set; CLI prints an incomplete-list banner; the well-formed test allows it on headless CI legs (where it immediately proved reachable). Legacy `list_audio_applications()` keeps best-effort behavior.

File: src/core/introspection.rs [hunk 11]
```diff
@@ -635,6 +718,65 @@ mod tests {
...
+    fn list_audio_applications_matches_scoped_applications() {
+        let plain = list_audio_applications().expect("infallible");
+        let scoped = list_audio_applications_scoped().expect("infallible");
+        assert_eq!(plain.len(), scoped.applications.len());
+        for (a, b) in plain.iter().zip(scoped.applications.iter()) {
+            assert_eq!(a.id, b.id);
+            assert_eq!(a.name, b.name);
+            assert_eq!(a.kind, b.kind);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (baladithyab) [reply src/core/introspection.rs:802]: ✅ fix-now (`984eeb2`): both tests now assert projection/well-formedness properties on ONE captured enumeration instead of comparing two live snapshots (the delegation is by-construction — `Ok(scoped()?.applications)` — so cross-snapshot equality only measured app churn).

File: bindings/rsac-python/src/lib.rs [hunk 12]
```diff
@@ -1495,6 +1502,771 @@ impl PyAudioCapture {
...
+                }
```
[the remark this answers is not in the document]
Comment (baladithyab) [reply bindings/rsac-python/src/lib.rs:2158]: ✅ fix-now: `59a5ae0` (as your re-check noted) — `Composition::__iter__` drops `iter_guard` before `start()`, mirroring `PyAudioCapture::__iter__` (`aaf678b`).

**Comment by baladithyab** [comment]
📌 Follow-up on the review-body nitpick (ExampleComposition hygiene, never an inline thread): fixed in #60 — `defer voice.Free()`/`defer system.Free()` right after construction and all `SetLayout`/`AddSource`/`SetSampleRate` errors checked instead of `_ =` discards.
